test/test_api/test_newtask_passwd.py

```diff
@@ -193,4 +195,173 @@ def test_newtask_no_gold_answers(self):
         res = self.app.get(url)
         assert res.status_code == 200, (res, res.data)
         task = json.loads(res.data)
-        assert task.get('gold_answers') is None
\ No newline at end of file
+        assert task.get('gold_answers') is None
+
+    # For task routing
+    @with_context
+    def test_newtask_with_task_preference_scores(self):
+        """Test newtask returns task with best matching score"""
+        user_info = dict(metadata={"profile": json.dumps({"english": 0.9})})
+        owner = UserFactory.create(id=500, info=user_info)
+        user_repo.save(owner)
+        project = ProjectFactory.create(owner=owner)
+        project.info['sched'] = Schedulers.user_pref
+        project_repo.save(project)
+
+        task_1_info = {'question': 'answer_1'}
+        task_1_pref = {'english': 1.0}
+        task_2_info = {'question': 'answer_2'}
+        task_2_pref = {'spanish': 1.0}
+        TaskFactory.create(project=project, info=task_1_info, priority_0=0, worker_pref=task_1_pref)
+        TaskFactory.create(project=project, info=task_2_info, priority_0=1.0, worker_pref=task_2_pref)
+        api_key = project.owner.api_key
+
+        # as a real user, no password
+        url = '/api/project/%s/newtask?api_key=%s' % (project.id, api_key)
+        res = self.app.get(url)
+        assert res.status_code == 200, (res, res.data)
+        task = json.loads(res.data)
+        assert task.get('info', {}).get('question') == 'answer_1'
+
+    @with_context
+    def test_newtask_with_task_preference_scores_invalid_value(self):
+        """Test newtask returns task with best matching score"""
+        user_info = dict(metadata={"profile": json.dumps({"english": "A"})})
+        owner = UserFactory.create(id=500, info=user_info)
+        user_repo.save(owner)
+        project = ProjectFactory.create(owner=owner)
+        project.info['sched'] = Schedulers.user_pref
+        project_repo.save(project)
+
+        task_1_info = {'question': 'answer_1'}
+        task_1_pref = {'english': 1.0}
+        task_2_info = {'question': 'answer_2'}
+        task_2_pref = {'spanish': 1.0}
+        TaskFactory.create(project=project, info=task_1_info, priority_0=0, worker_pref=task_1_pref)
+        TaskFactory.create(project=project, info=task_2_info, priority_0=1.0, worker_pref=task_2_pref)
+
+        api_key = project.owner.api_key
+
+        # as a real user, no password
+        url = '/api/project/%s/newtask?api_key=%s' % (project.id, api_key)
+        res = self.app.get(url)
+        assert res.status_code == 200, (res, res.data)
+        task = json.loads(res.data)
+        assert task.get('info', {}).get('question') == 'answer_2'
+
+    @with_context
+    def test_newtask_with_task_filter_1(self):
+        """Test newtask returns task that matches filter"""
```

Comment:
Looks like that there's scope of explaining in details what the test does under docstring here and under functions below(_2, _3, _4) as they seems to be similar right now.